### ISSUE TITLE
Draw Batches in Epochs

### ISSUE THREAD

**Issue by pl-fuchs** [tummfm/jax-sgmc#9]
Random data loading should allow to draw batches in epochs, such that each observation is drawn exactly once per epoch.
To maintain a correct shape the last batch returned might contain partially invalid samples. Those are marked via a masking array in the `MiniBatchInformation` tuple.

For example for the simple dataset with 10 observations

```python

x = onp.arange(10)
data_loader = NumpyDataLoader(x=x)
init_fn, batch_fn = data.random_reference_data(data_loader, 2, 3)

```

a different method is passed for the second chain:

```python 

random_chain = init_fn()
epoch_chain = init_fn(shuffle=True, in_epochs=True)

```

The default chain always returns valid samples, but the epoch-wise chain returns invalid samples every fourth period:
```python

def eval_fn(chain):
  for _ in range(4):
    chain, batch = batch_fn(chain, information=True)
  print(batch)

eval_fn(random_chain)
  # ({'x': DeviceArray([4, 6, 6], dtype=int32)}, MiniBatchInformation(observation_count=10, mask=DeviceArray([ True,  True,  True], dtype=bool), batch_size=3))
eval_fn(epoch_chain)
  # ({'x': DeviceArray([5, 0, 0], dtype=int32)}, MiniBatchInformation(observation_count=10, mask=DeviceArray([ True, False, False], dtype=bool), batch_size=3))
```
